File: security-application/src/main/java/com/example/security/application/query/StatsResult.java
package com.example.security.application.query;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * 시계열 / 집계 결과.
 *
 * @param series 시계열 — bucket 시각별 카운트 (groupByField 없는 경우)
 * @param topGroups groupByField 가 있을 때만 채워짐 — 그룹 키 → bucket 시각별 카운트
 */
public record StatsResult(
    List<TimeBucket> series, Map<String, List<TimeBucket>> topGroups) {

  public StatsResult {
    Objects.requireNonNull(series);
    series = List.copyOf(series);
    topGroups = topGroups == null ? Map.of() : Map.copyOf(topGroups);
  }

  public record TimeBucket(Instant timestamp, long count, double p95LatencyMs) {}

  public static StatsResult empty() {
    return new StatsResult(List.of(), Map.of());
  }
}

File: security-application/src/main/kotlin/com/example/security/application/query/StatsResult.kt
package com.example.security.application.query

import java.time.Instant

/**
 * 시계열 / 집계 결과.
 *
 * @property series 시계열 — bucket 시각별 카운트 (groupByField 없는 경우)
 * @property topGroups groupByField 가 있을 때만 채워짐 — 그룹 키 → bucket 시각별 카운트
 */
@JvmRecord
data class StatsResult(
    val series: List<TimeBucket>,
    val topGroups: Map<String, List<TimeBucket>>,
) {

    @JvmRecord
    data class TimeBucket(val timestamp: Instant, val count: Long, val p95LatencyMs: Double)

    companion object {
        @JvmStatic
        fun empty(): StatsResult = StatsResult(emptyList(), emptyMap())
    }
}
